Syndication/AzureStack.MarketplaceSyndication.psm1

```diff
@@ -90,165 +83,1086 @@ function Sync-AzSOfflineMarketplaceItem {
+                        $checktool= Test-Path "C:\Program Files (x86)\Microsoft SDKs\Azure\AzCopy\AzCopy.exe"
+                        If ($checktool -eq $true){
+                            DownloadMarketplaceProduct -Source $zipsource -Destination $zipdestination -ProductName "$zipName.zip" -PremiumDownload -MaxRetry 2
+                            "$productFolder\$zipName.zip"|out-file "$productFolder\$azpkgName.txt" -Append
+                            $productDetailsProperties['sourceBlob'].uri = "$productFolder\$zipName.zip"
+                        }
+                        else{
+                            Write-Host "Please install Azure Storage Tools first,canceling"
+                        }
+                    }else{
+                        DownloadMarketplaceProduct -Source $zipsource -Destination $zipdestination -ProductName "$zipName.zip" -MaxRetry 2
+                        "$productFolder\$zipName.zip"|out-file "$productFolder\$azpkgName.txt" -Append
+                        $productDetailsProperties['sourceBlob'].uri = "$productFolder\$zipName.zip"
                     }
                 }
-                'virtualMachineExtension' {
-                    # download zip
-                    $zipName = $productDetails.properties.galleryItemIdentity
-                    $zipsource = $downloadDetails.properties.sourceBlob.uri
+            }
+
+            {($_ -eq 'resourceProvider') -or ($_ -eq 'solution')} {
+                # download zip
+                foreach ($container in $downloadDetails.properties.fileContainers)
+                {
+                    $zipsource = $container.sourceUri
+                    $containerName = $container.id
                     If ([string]::IsNullOrEmpty($zipsource)) {exit} else {
-                        $FileExists = Test-Path "$destination\$zipName.zip" 
-                        If ($FileExists -eq $true) {Remove-Item "$destination\$zipName.zip" -force} else {
-                            New-Item "$destination\$zipName.zip" 
+                        $zipDestination = "$productFolder\$containerName"
+                        if ($container.type -match 'zip'){
+                            $zipDestination = "$productFolder\$containerName.zip"
+                        } else {
+                            $zipuri = [uri]$zipsource
+                            $fileName = $zipuri.Segments[-1]
+                            $dotposition = $fileName.LastIndexOf('.')
+                            if ($dotposition -ne -1)
+                            {
+                                $fileExtension = $fileName.Substring($dotposition+1)
+                                $zipDestination = "$productFolder\$containerName.$fileExtension"
+                            }
+                        }
+
+                        $FileExists = Test-Path $zipDestination
+                        If ($FileExists -eq $true) {Remove-Item $zipDestination -force | Out-Null}
```

Comment:
Is this the right behavior? Just delete and redownload every time? If I rerun script I probably expect it to no-op on what is already download. Can you clarify scenario if this makes sense?

Reply:
Current behavior is to redownload it. One thought is if previous download somehow left something half downloaded, maybe easier to redownload it. I agree we can have more logic here to detect and skip already downloaded items. Will look at it later.